### ISSUE TITLE
Support for .NET Core

### ISSUE THREAD

**Issue by jonorossi** [castleproject/Windsor#145]
No one active on the Castle team is looking at this so I thought I'd log an issue to see if anyone wants to contribute to this work. It should follow the way we've done things with Castle Core.

**Comment by BMarques** [comment]
I forked the project just so I have "fun" to add support for .NET Core, No promises I'll ever finish doing it, but it's a good exercise to learn more .NET features.

Before doing anything and because Castle Windsor will be upped to version 4, I was wondering if it's a good idea to remove support for Silverlight and old versions of .NET, and the oldest .NET supported would be 4.5.2.

From what I saw the most common error is the Serializable attribute, which is not supported in .NET Core. Did this happen with Castle Core project? What's the recommended plan of action ? Create an attribute to do the same as previously or do something else?

**Comment by jonorossi** [comment]
@BMarques yes, we'll drop Silverlight support here too (no choice, I removed it in Castle Core 😄 ), however the Silverlight conditional compilation was useful while doing the port, so just remove the SL project from the solution in your fork and don't worry about it.

.NET Framework versions is a different story. I decided I wanted to keep older versions for Castle Core to allow the mocking libraries to support older frameworks, however we don't need to do that here so for now just ignore older frameworks in your fork. We can make a decision in the future if we need older frameworks.

Yes, `SerializableAttribute` came up in the Castle Core port. I decided not to shim a fake attribute in as it functionally does nothing, however it looks like the attribute will be back by the end of the year in .NET Core so do whatever you need to get something going. I originally did two PRs on Castle Core master for serialization (attributes and other code) and reflection (deprecation of old APIs) to simplify our netcore branch, however it sounds like all the old APIs are coming back soon.

I'd recommend taking a look at the actual Castle Core code, especially the list of conditional compilation symbols in the README we've defined to opt-in functionality for .NET Framework.

**Comment by BMarques** [comment]
After you mentioned that Serializable will be back later, I saw that indeed .NET Core 1.1 will be closer to .NET Framework, so I think for the moment I'll wait and use .NET Framework 4.6.1 and using Castle.Windsor.MsDependencyInjection

**Comment by jonorossi** [comment]
@BMarques no problem. If you weren't already aware, note that the Castle.Windsor.MsDependencyInjection package isn't published by the Castle Project.

**Comment by paragmeshram** [comment]
Excited to see .NET Core support as Windsor is most popular, matured and robust IoC with many more other capabilities

**Comment by alinapopa** [comment]
Hi @jonorossi ,
I'd like to help with the porting to .NET Core. I see that in Castle.Core the old projects/solution were kept and a new solution and projects was created for .NET Core; do you plan the same for Castle.Windsor?

**Comment by jonorossi** [comment]
@alinapopa we did that for Castle Core because the .NET Core tooling was in a massive state of flux back then when it was still really early DNX builds, however there is no requirement to do it that way. Now that .NET Core is going back to MSBuild I'd actually like to see a single set of csproj files.

I don't see any problem with Windsor being .NET Standard 2.0 either if that makes it heaps easier, if we have to conditionally compile code out I'd like us to use feature-based symbols (like we've done with Castle Core) so it is clear when reading the code why the code can't work. For example, `FEATURE_REMOTING` or `FEATURE_WCF`.

It is up to you, but I think a good first step would be to get a build script set up (maybe like our Castle Core one) that we can set up on TeamCity even if the .NET Core build fails compilation, then submit smaller PRs working towards a working build rather than one giant PR which really slowed things down last time and made code review difficult.

This is what TeamCity is currently running for the .NET Core build for Castle Core:
```
$ buildscripts\Set-VersionInfo.ps1
$ build NETCORE
```

Let me know if you've got any questions, feel free to submit a WIP pull request and we can discuss things on the code. Nothing is set in stone, don't be afraid to ask any questions or make any suggestions on how to do things, there is heaps of history here but that doesn't mean those opinions are still valid.

**Comment by alinapopa** [comment]
@jonorossi having a single set of csproj files means the solution can only be opened with Visual Studio 2017 RC. The existing csproj files need to have many changes, it not just adding a new target framework. It's probably better at this time to do the porting in the same manner as it was done for Castle.Core.
Using .NET Standard 2.0 would make things significantly easier, but the test project which needs to use dotnet-test-nunit is not compatible at this time.

**Comment by Kralizek** [comment]
Are you guys going to target .NET Core 1.1 or 1.0?
If possible, I'd suggest 1.0 because it's the current LTS, whilst 1.1 is just Current.
For example AWS won't use Current releases of .NET Core runtime, but only the LTS ones.

**Comment by jonorossi** [comment]
@alinapopa sounds like a good way forward. Once all the tooling settles with a VS RTM we'll have a better idea of how things need to work and will hopefully be able to consolidate the project files here and in Castle Core at that point.

@Kralizek what do you mean by "AWS won't use current releases", can't you just deploy your own version of .NET Core to EC2? What AWS service are you referring to?

**Comment by Kralizek** [comment]
@jonorossi I'm referring to AWS Lambda. They now support C# by leveraging .NET Core runtime. Specifically, they will only use LTS versions of it.

**Comment by fir3pho3nixx** [comment]
Hi All, 

I have been working with @jonorossi to move us closer to making this happen. Not having Windsor on Core with Linux quite simply does not feel right :)

We are actively clearing out the technical facets of legacy frameworks so we can better position ourselves to leap to core(whilst minimising the distance). The goal right now is to get a build going. 

#What have we done so far?

 - Removed build support for SILVERLIGHT, NET35, NET40, CLIENTPROFILE and MONO. 

#What do we need?

 - Remove build flags for the above(please don't remove SILVERLIGHT!). I have started on NET35 @ PR #195. 
 - Create build infrastructure for NETCORE intently observable @ [The Castle Build Server](http://builds.castleproject.org/project.html?projectId=Windsor&tab=projectOverview) (please register).
- Make build pass(hopefully with first class Linux & Mac support, @jonorossi is this possible?)!
- Supporting AWS Lambda is a 'must have' for me. My team is crossing over from Azure Fucktions(good shout @Kralizek). 

@alinapopa - I have had a look at your WIP #160. There are some great learning's there you could help us with. I must say doing it all in one hit is a bit too much change in one go. It also means @jonorossi has to keep polling the PR for commits. We *need* to do this using baby steps by raising multiple PR's whilst keeping an eye on tests. This will help @jonorossi because he will get a notification each time we raise a PR without having to monitor things manually. To give you an idea, I first forked the project and used a 'brute force' approach to get Windsor compiling on Core just to understand the problems involved. I can honestly say I would not merge this in one hit. I urge you to rejoin this thread :)

**Comment by jonorossi** [comment]
> Make build pass(hopefully with first class Linux & Mac support, @jonorossi is this possible?)!

Linux is easily do-able, the Castle Core builds on TeamCity __currently__ run against Mono however the .NET Core one only tests against Windows. macOS is harder because it can only run on Apple hardware, so the cost of getting a macOS machine in the cloud is very expensive.

**Comment by paragmeshram** [comment]
Hi all,

I have seen a lot of traction in recent weeks and I would like to take a moment to applaud you guys for the time and efforts you have been putting. No doubt that Castle Core & Windsor are most matured open source projects I have seen. Long live the Castle !!!

Cheers,
Parag Meshram

**Comment by fir3pho3nixx** [comment]
OK, NET35, NET40, CLIENTPROFILE and MONO are gonners. We are keeping SILVERLIGHT build flags(not build process) because this is going to be very useful going forward.

**Comment by fir3pho3nixx** [comment]
@jonorossi - How do you want to tackle the project.json issue?

To quote @MaximRouiller from https://github.com/aspnet/Home/issues/1433

![image](https://cloud.githubusercontent.com/assets/607402/23628366/f54de482-02ab-11e7-84ad-8b0f49ad1789.png)

**Comment by jonorossi** [comment]
> How do you want to tackle the project.json issue?

How do you feel about skipping over project.json and using VS2017 RC and the new csproj tooling especially since VS2017 should RTM today?

**Comment by alinapopa** [comment]
@jonorossi @fir3pho3nixx please let me know how I can help. For smaller PR's, are you thinking of something like splitting the PR in smaller ones, or having an entirely different approach? Also, what would be a reasonable size?
I vote for using csproj and VS2017.

**Comment by jonorossi** [comment]
@alinapopa it is not so much the size of the PR but the number of different somewhat unrelated changes all at once. It is important that we don't inadvertently introduce regressions because things were overlooked with a massive PR where it can sometimes be hard to know the intent/reason behind each change and to know if there could be better ways to do things.

I think the first step should be to add the new csproj and get a CI build going even if it doesn't compile. We can then start getting some of the reflection usages updated so both platforms compile (maybe GetTypeInfo as one PR, other PRs can have more than one reflection method change), then keep stepping in the direction to reduce compilation errors.

I'm not directly working on the code so it is best to decide with @fir3pho3nixx who is going to do what.

**Comment by fir3pho3nixx** [comment]
@alinapopa - Welcome back! Start by getting the project setup going, I still have bits and bobs to take care of with the PR's for removing the build flags(also have to install VS Next). It does not matter if the build does not pass just yet for Core. As long as the build/tests pass for the NET45 related stuff. 

@jonorossi - Should we do it as a separate solution just like in Castle.Core?

**Comment by MaximRouiller** [comment]
"Shits on fire, yo" guy reporting in for duty.

Don't waste time on `project.json`. It was loved. It was fun. It ended up dead in a ditch.

**Comment by jonorossi** [comment]
> Should we do it as a separate solution just like in Castle.Core?

To be honest I don't know, I suspect we'll probably want to get rid of the current csproj files that specify each file, especially since the new tools can target .NET 4.5 via .NET Standard, and we aren't targeting those old frameworks anymore. I'd be interested to know if we could get rid of all those MSBuild customisations and the targets file that hooks in.

For now though it might be best to create another project side by side until it is building and we can verify all the functionality of the old build scripts is migrated/obsoleted.

**Comment by hikalkan** [comment]
I was about to leave Caslte Windsor because of not supporting .netstandard (even I love it), but I see that you are working on it. Do you have any time estimation, at least for a beta?
Thanks a lot.

**Comment by jonorossi** [comment]
> I was about to leave Caslte Windsor because of not supporting .netstandard (even I love it), but I see that you are working on it. Do you have any time estimation, at least for a beta?

@hikalkan it really comes down to contributions, @fir3pho3nixx is currently the only one making any progress. Not to sound rude, but if you've got the time to change to another library than I would hope you could contribute some small amount of time to making this happen instead.

**Comment by hikalkan** [comment]
I really would want to do it. It would be better for me too. Do you have an idea where I can start?

**Comment by jonorossi** [comment]
@hikalkan I'll let @fir3pho3nixx respond as I don't know if he is already working on it at the moment, but the next thing is to get in new project files so everyone can start using the new tooling.

**Comment by hikalkan** [comment]
I just moved ABP framework to VS2017. But don't know if Castle solution is more complicated (since it supports multiple targets). @fir3pho3nixx can I help to the project and how?

**Comment by hikalkan** [comment]
It's hard me to make such a significiant solution structure change without understanding the overall structure. The solution has it's own build system. Even if I can make it somehow working, it will be not the way you prefer.

One another offer that we can be a sponsor for this development (as Volosoft) if one of main contributors of Windsor can handle it. We can discuss details of this offer via email (info@volosoft.com).

**Comment by fir3pho3nixx** [comment]
@hikalkan - Feel your pain! Thanks for the offer of sponsorship would take it if I was in between contracts but alas I am fully engaged at the moment, so my issue is simply not having enough time. That said I do believe we are weeks away from making this happen. Not months. There has been considerable amount of time spent cleaning things up. I think we are finally in a position to start crunching out some 'hadouken' PR's to move the project closer to core. As a courtesy will probably @mention everyone on this thread once we are done. :)

![ryuhadouken](https://cloud.githubusercontent.com/assets/607402/23918968/e664ce1c-08ec-11e7-93d4-f6115243af91.png)

**Comment by hikalkan** [comment]
Thank you @fir3pho3nixx, it's very appreciated. I like Castle Windsor much and want that it continue to it's life in .net core too. I'm open for sponsorship and contribution. I would want to contribute code but it's hard since you are changing solution structure. But please write me if anything I can do for it.
Thanks.

**Comment by fir3pho3nixx** [comment]
@hikalkan - I can see you ran into issues getting Castle.Core to instantiate Dictionaries properly using Castle.Core v4. Think I might have fixed this issue. Also have castle windsor targeting all Desktop CLR versions using multi targeting(net45, net451,net461,net462) using VS2017. Hope to be moving to core soon. Alas I don't think the guys at Castle are willing to take this change. It is freaking huge. It also means these guys have to delete alot of code to make this work which they are not prepared to do right now. 

Might create a fork and publish a new NuGet still honoring the existing Apache license by keeping the copyright headers but think I might call it "Fortress.Windsor". 

Will keep you updated and pass anything I can back to the Castle team if I can. It is hard, there is a alot of cruft in this project.

**Comment by hikalkan** [comment]
Thank you @fir3pho3nixx. I also think to create a new fork would be good. Don't hesitate to contact me when you need. Our team want to help on that if we can.

**Comment by jonorossi** [comment]
@fir3pho3nixx I don't know what happened over the last 4 days but it seems you have hastily decided to close all of your open pull requests and issues, delete some of your comments, and create your own fork.

* #196
* #205 - deleted comments
* #211
* #212 
* #215
* https://github.com/castleproject/Attic/pull/2

I really don't understand why, but it seems you took me not responding at your beck and call on Thursday/Friday as me ignoring you. I really don't think I needed to explain but Thursday/Friday for me I was doing paid work, I didn't even get a chance to even look through Castle emails.

> Alas I don't think the guys at Castle are willing to take this change. It is freaking huge. It also means these guys have to delete alot of code to make this work which they are not prepared to do right now.

Please stop making stuff up, if I wasn't interested in the work why would I dedicate my time to discussing it, reviewing and merging pull requests.

> Might create a fork and publish a new NuGet still honoring the existing Apache license by keeping the copyright headers but think I might call it "Fortress.Windsor".

There is no need for snide comments towards the Castle Project, it looks like you've already begin your fork: https://github.com/cryosharp/fortress.

> Will keep you updated and pass anything I can back to the Castle team if I can. It is hard, there is a alot of cruft in this project.

Yes there is a lot of cruft, which is why things need to be worked through in a calculated manner to ensure we don't disappoint users that have been using Windsor for a long time and built up their projects with it.

I've also responded to some of your comments in https://github.com/castleproject/Windsor/issues/214#issuecomment-287676591 and noted your breach of our code of conduct. If you wish to continue engaging with the Castle Project please respect our code of conduct.

**Comment by fir3pho3nixx** [comment]
Apologies for the conduct it is merely frustration. This will move quicker on my fork. The turn around time for the PR's and knowing how those should be structured without a member doing setting up a new build on the new tooling leaves contributors(or potential contributors) guessing at what changes they should be helping out with. This is a big problem for guys willing to help from the outside.

**Comment by fir3pho3nixx** [comment]
This change is not small and I don't it can be done by a contributor by way of a PR. It has to be setup by a member with access to the build infrastructure and possibly by merging core and windsor together.

**Comment by Kralizek** [comment]
Maybe a working PR that breaks the build chain would be a good starting point for repo curators?

**Comment by fir3pho3nixx** [comment]
The only way I could get this done was by deleting the *.csproj's and all the MsBuild stuff. Then laying down the new VS2017 msbuild/csproj structure. Everything had to be re-referenced, so I though I would do the right thing and pull new packages where I could. 

1. NUnit latest breaking changes bit me. So the project would no longer compile. 
2. Feature based build flags needed to be put back. Things started compiling again. Careful thought need to be given to what is deprecated. Accidentally putting this in PR's will cause additional round trips which take time.
3. Tests started failing because of method access violations with the latest verison of Castle.Core(which means the security attributes needed some thought). I pulled Castle.Core into the Windsor project.
4. There was also alot of faffing around with embedded resources and things to get the existing tests to run.  
5. You lose the ability to run and debug tests using Resharper. So then created a VS2015 side by side solution for that. It helped fix alot of things but added more weight to the change set. 
6. Enabling tests for multiple Desktop CLR versions started highlighting issues with PEVerify and the saving of assemblies in Castle.Core. So ModuleScope in DynamicProxy needs a bit of work to get round this because of I/O colisions around the saving of assemblies whilst running tests for let's say net45 and net451 as an example. It also highlights the need for more changes in Castle.Core. 
7. Not having admin access to teamcity or using 'experimental branches' outside of master left me dangling a little as to how we get a green(DesktopClr)/red(CoreClr) build going. 

This is all the stuff you need just to get multi-targetting(net45,net461) going on new VS2017 tooling before you even start thinking about dropping framework monikers for dotnet standard or core. 

@jonorossi - Again I do appologise my frustrations got the better of me. I still believe this is bigger than anyone single contributor, it needs the weight of a member behind it to help drive things forward. Some of the PR's also depend on each other, and run in a chain, so every now and then I find myself going into holding pattern before I can pull them into the master branch on my fork and continue. Also working ahead is difficult because if one of them don't get merged it gives one the feeling that the rest wont make it in or take more time to get there holding things up even more.  

@Kralizek - I agree but these guys are busy and @jonorossi is the only one at the moment making time for this which is sparse. An example is #205 which has been open for 10 days now. I pleaded with @jonorossi to expedite the removal of this because grepping across the project was causing alot of noise(that is what was in the deleted comments), not merging things like makes it hard to move when taking something like this work on. 

It is big work and a slow turning process.

**Comment by fir3pho3nixx** [comment]
Grrr... Had another fun evening working on ModuleScope in Castle.Core to try and parallelise tests for multi framework targeting. The trade off here is we need to drop off friend assemblies as a feature in castle dynamic proxy. Microsoft simply does not support AOP based `InternalsVisibleTo` attributes for dynamically named `strong named` assemblies. This is entirely possible if we plan to use C++ but that goes to a different place entirely. 

@jonorossi - We can cuss each other out as much as we like. It helps none. You need help, you need to steer my energy or I am going it alone. If this earns me a ban from Castle, so be it. I wont stop until this project is on Core. With or without you.

**Comment by fir3pho3nixx** [comment]
Just to set the record straight #214 was updated with responses.

**Comment by fir3pho3nixx** [comment]
Just for record I updated #214.

**Comment by fir3pho3nixx** [comment]
Anyone interested in a dotnet core version of Castle Windsor ... 

Please go here: https://www.nuget.org/packages/Fortress.Windsor/

Let's hope this does not last long!

**Comment by fir3pho3nixx** [comment]
@jonorossi  - think it is time we have a call ... email me using my profile.

**Comment by jonorossi** [comment]
> think it is time we have a call ... email me using my profile.

I'd be happy to work with you to plan this and get things going, but I checked out for a break after being told I've apparently got a God complex and my country of birth being brought into the discussion just a day after your apology. I'd be happy to draw a line in the sand, and try to forget the past so we can move forward.

At this stage I'm not interested in a phone call, however if you'd like to begin an email conversation on how you think we could/should proceed from here my email is "jono" at my username .com.

**Comment by fir3pho3nixx** [comment]
All good! When do we start?

**Comment by fir3pho3nixx** [comment]
@kkozmic - What are your thoughts on `CallContextLifetimeScope`. This is the foundation of creating scope accessors which hangs off LifestyleType.Scoped. This is very handy stuff for managing(brain dead easy) life styles using the Windsor API in core. NetStandard2.0 might help us out but do you think we should create a new implementation or wait?

Busy writing a custom life style manager ATM(your adherence to the open/closed principle is paying dividends).

**Comment by generik0** [comment]
@fir3pho3nixx and @jonorossi from a long time user of Castle Windsor it is really Great to see your work...
On the orker hand it is sad to see you discussion and that you have hit a conflict. Normally when this happens is just because people are stressed. And the only thing that comes out of it, is hurt feelings, on both sides.

We all hope you work it out and continue your wonderful work. I am sure the community has patience for this release. Don't stress about us!

I have a casual development business on the side. I would also like to contribute. As Castle has made my life easier and better. Mail me, let me know how I can sponsor something...
I would love to help with some code and tests. But I don't think I am good enough....

- G

**Comment by fir3pho3nixx** [comment]
Encouraging words. I have been exerting a lot of pressure on @jonorossi. I hope to change this going forward(I will openly admit I have been out of order). I just wish we were leveraging the community better, all people from all backgrounds coming together to achieve a great result. This is not happening right now for whatever reason. People are willing to put effort in (significant hero efforts, @alinapopa being a case in point). @jonorossi has a big task in front of him trying to steer this. I have build engineering suggestions which will enable github status checks (where tests run on PR's and proposed merge's to target branches across windows, mac, linux). Hopefully this will encourage guys like yourself to have a go at contributing. Tests will run and you will immediately know whether you got it right or not. This will also give @jonorossi confidence whilst reviewing/merging PR's, if tests fail it goes straight back to the contributor. My position now is I want a conciliatory approach to this. I am done apologising, yes I am angry about this and if this is going to turn into a shaming contest I am out.

**Comment by generik0** [comment]
Yes, contributing would be cool... Test are great. Then people like me could feel more safe in contributing, as you say...!

I don't think anyone should be shamed, I would recommend everybody just say F@&k 'it, move on and focus on what lies ahead :-)

**Comment by Kralizek** [comment]
Changing topic. :)
Shall we try to list the area of intervention? This will make easier to understand the amount of work needed.

I don't know so much about the internals of the project, but one extra challenge in the .NET Core/Standard space is the cooperation with `IServiceCollection` and `IServiceProvider`.

I guess the container/kernel should be able to ingest registrations done on the first one and we should offer a adapter of the container towards the second one.

**Comment by hikalkan** [comment]
Hi,

Actually, moving to .netstandard/.netcore is different than IServiceCollection work (https://www.nuget.org/packages/Microsoft.Extensions.DependencyInjection package integration). It's necessary to make Windsor usable in AspNet Core applications. But without this integration, moving to .netcore is not meaningful in pratical since most of developers want to use this in AspNet Core.

Good new is that; I have already created an adapter for it: https://github.com/volosoft/castle-windsor-ms-adapter/ and working on it to make it compatible with @fir3pho3nixx's work. But if an official adapter will be released, I will be happy to drop my project :)

**Comment by Kralizek** [comment]
@hikalkan I agree with you. Not required to perform a successful migration to .NET Standard. Required to deliver a usable/used product :)

**Comment by jonorossi** [comment]
For those that don't follow the development mailing list, I just wrote some important info:
https://groups.google.com/d/topic/castle-project-devel/vC8EO1r2d64/discussion

**Comment by jeremymeng** [comment]
@jonorossi #220 is close to completion and @alinapopa starts looking at porting to .NET Core again.  Do you still prefer multiple small pull requests, or big pull requests with clearly separated and independent commits? I agree that small pull requests are easier to review. At the same time they also add the overhead of branching/re-basing. And each PR might not result in a good build.

Either way, we will go with your preference.

**Comment by generik0** [comment]
Hi
Thisted be a stupid question. But this issue is support for .net core.
Shouldnt it be implement with .net standard?

**Comment by alinapopa** [comment]
Yes we will target .NET Standard. .NET Standard is a specification of what API's to be available, and .NET Core implements .NET Standard.
https://docs.microsoft.com/en-us/dotnet/articles/standard/library

**Comment by jonorossi** [comment]
> Do you still prefer multiple small pull requests, or big pull requests with clearly separated and independent commits? I agree that small pull requests are easier to review. At the same time they also add the overhead of branching/re-basing. And each PR might not result in a good build.

I'll work with as small or big of a pull request you guys want to submit, and now that I'm back after having my head down in work for a few days I'll get to them as quickly as possible. With that said try to keep pull requests complete and whole, otherwise that is what makes it really hard to follow when there is so much back and forth on a big pull request.

**Comment by Kralizek** [comment]
I'd like to contribute in my spare time. Anything I can help with?

**Comment by fir3pho3nixx** [comment]
@Kralizek - You can definitely help. 

I am busy migrating Castle Core over to using AppVeyor and TravisCI here: https://github.com/castleproject/Core/issues/241. Nearly done. 

@alinapopa and @jeremymeng  are busy looking at the same thing over here: 
https://github.com/castleproject/Windsor/issues/220. They still have a couple of PR's left. 

What we need to start thinking about is how we apply the "feature" based build symbols that would enable Castle.Windsor to target netstandard. For example netstandard does not support app domains currently(although this might be coming back in netstandard2.0),  so in Castle Core conditional compilation symbols are applied depending on the particular target framework moniker that the source is being compiled for. The same thing has to be done in Castle Windsor. 

If you fork Castle Core and checkout the `refactor-build` branch, the compilation symbols are defined for each moniker in the VS2017 tooling here: https://github.com/castleproject/Core/blob/refactor-build/buildscripts/common.props#L21. They are also documented here: https://github.com/castleproject/Core/tree/refactor-build#conditional-compilation-symbols. I would use `FEATURE_APPDOMAIN` as an example. Then start looking into the Castle Windsor code to see where there are opportunities for doing the same thing. Once you are familiar with how this works, ping back on this issue with your findings/thoughts. We can then further collectively discuss how you go about raising your first pull request on Castle Windsor and help you with any questions. 

@jonorossi - Do you think this would be a good approach?

**Comment by jonorossi** [comment]
> Do you think this would be a good approach?

Sounds reasonable to learn how things are set up.

It is probably best to do either `FEATURE_SERIALIZATION` or `FEATURE_LEGACY_REFLECTION_API` first as they'll knock out most of the compilation errors on .NET Standard, however @alinapopa has already done most of them in her pull request (#160), so it is probably best to wait a little bit so we can get that in.

**Comment by alinapopa** [comment]
I created this checklist to track the work needed:
- [ ] Cleanup of the old build setup.
- [ ] Update Castle.Core dependency to version 4.0.
- [ ] Add .editorconfig
- [ ] Add conditional compilation for features only in .NET Framework, like Castle.Core
    - [ ] `FEATURE_APPDOMAIN`
    - [ ] `FEATURE_SERIALIZATION`
    - [ ] `FEATURE_SECURITY_PERMISSIONS` 
    - [ ] `FEATURE_REMOTING`
    - [ ] `FEATURE_SYSTEM_CONFIGURATION`
- [ ] Port code with alternatives in .NET Core
    - [ ] Use Type.GetTypeInfo() to replace the old Type methods
    - [ ] Replace [MethodImpl(MethodImplOptions.Synchronized)] with locks
    - [ ] System.Web and System.Web UI: PerWebRequestLifestyle is for use with ASP.NET. Remove it on .NET Standard.
    - [ ] remove the use of ICloneable 
    - [ ] Replace System.Array methods with ones that exist on Core
    - [ ] Get alternatives for Assembly.GetCallingAssembly: replace with a method that requires the assembly name to be passed in
    - [ ] Replace Delegate.CreateDelegate with MethodInfo.CreateDelegate
    - [ ] Use AssemblyLoadContext for dynamic loading of assemblies
    - [ ] Windows perf counters: Consider replacing with Diagnostic.Tracing.EventCounter (https://github.com/dotnet/corefx/issues/3906 ) or remove.  
- [ ] Port tests and fix test failures
- [ ] Cleanup Silverlight support
- [ ] Setup build for .NET Core

**Comment by fir3pho3nixx** [comment]
@alinapopa - that is an awesome start for creating a [Project Board](https://help.github.com/articles/creating-a-project-board/)

**Comment by jeremymeng** [comment]
@fir3pho3nixx [There's one already](https://github.com/castleproject/Windsor/projects/1), It seems read-only to us though.

**Comment by fir3pho3nixx** [comment]
@jonorossi - can you give @alinapopa write access to this board? We also need more columns to track status like "Planned", "Ready", "In Progress", "Done". We can also track who is doing what with this.

**Comment by jonorossi** [comment]
> can you give @alinapopa write access to this board?

I tried working this out when I created that board but it appears I have to grant write access (https://help.github.com/articles/repository-permission-levels-for-an-organization/) to the repository to allow this. I've now granted write access to Windsor to @fir3pho3nixx and @alinapopa, I've set up master as a protected branch.

I'd love to get the remaining build refactor done and get this branch merged. Because of how the new .NET tooling works there are now quite a few people that need to use Windsor with Castle.Core 4.0 to use mocking libraries: #227. I suspect the .NET Core work will take a few weeks to sort out so it'd be good to get a Windsor 3.5.0 release out now.

**Comment by alinapopa** [comment]
@jonorossi PR #268 is the last one to get the main Castle.Windsor project to build on .NET Standard. Castle.Facilities.EventWiring and Castle.Facilities.FactorySupport projects can be easily ported. 
Castle.Facilities.Logging project has heavy dependency on `Castle.Services.Logging.Log4netIntegration` and `Castle.Services.Logging.NLogIntegration` which are not available for .NET Standard.
Castle.Facilities.Synchronize project has dependencies on Windows UI, I assume it doesn't make sense to port it to Core.
The tests would need to be split in separate projects, since the tests that use Castle.Facilities.Logging  and Castle.Facilities.Synchronize  will not be built for .Net Core.

**Comment by jonnii** [comment]
@alinapopa 🙏 👏

**Comment by jonorossi** [comment]
> PR #268 is the last one to get the main Castle.Windsor project to build on .NET Standard.

#268 merged 🎉 

> Castle.Facilities.EventWiring and Castle.Facilities.FactorySupport projects can be easily ported.

👍 

> Castle.Facilities.Logging project has heavy dependency on Castle.Services.Logging.Log4netIntegration and Castle.Services.Logging.NLogIntegration which are not available for .NET Standard.

log4net 2.0.8 supports .NET Standard however we haven't got around to fixing the Castle integration code since APIs are missing in the .NET Standard build (https://github.com/castleproject/Core/issues/201). 

NLog 5 is still in beta so we can't add support yet (https://github.com/castleproject/Core/issues/200).

The logging facility shouldn't have a heavy dependence on those, other than some unit tests I'd have only expected the `LoggerImplementation` enum and the weak reference with the assembly types in `LoggingFacility`. @fir3pho3nixx definitely agrees with me that all this logging stuff is overly complicated, he is just more vocal 😉 , however let's try to get what we can get building for now and we can redesign things in a future release.

> Castle.Facilities.Synchronize project has dependencies on Windows UI, I assume it doesn't make sense to port it to Core.

Yep, no need on .NET Core.

> The tests would need to be split in separate projects, since the tests that use Castle.Facilities.Logging and Castle.Facilities.Synchronize will not be built for .Net Core.

I think we'd be better off just using `DOTNET45` to exclude the unit tests for those facilities for now, I can see us simplifying things and depreciating some more facilities in the medium term.

**Comment by fir3pho3nixx** [comment]
@alinapopa - top work!!! :)

If you want we can steal the logging facility from fortress. @hikalkin helped me port this.

**Comment by Kralizek** [comment]
@alinapopa great job! :)

About the logging, why don't we migrate so that we use `Microsoft.Extensions.Logging` infrastructure? It's pretty well done and it lets the final user to decide which logging facility to use.

**Comment by jonorossi** [comment]
> About the logging, why don't we migrate so that we use Microsoft.Extensions.Logging infrastructure? It's pretty well done and it lets the final user to decide which logging facility to use.

@Kralizek I'd rather prefer we just not port the current logging facility to .NET Core and get this release out at this stage. There are thousands of package downloads for the current logging facility for .NET Framework so we can't just drop it, I think it would be best not to hold up getting the Windsor core out with discussions about or thrash with a new logging facility.

**Comment by Kralizek** [comment]
Fair enough. My suggestion was to spare some migration work and to take advantage of the major version change. But i see your point :)

**Comment by fir3pho3nixx** [comment]
@Kralizek - We have an issue raised by @alinapopa  on the project board @ https://github.com/castleproject/Windsor/projects/1 which is to do with removing SILVERLIGHT build symbols. Want to have a go?

**Comment by Kralizek** [comment]
@fir3pho3nixx I'm away until the 18th of June. I might give a look at it if it's still available when I'm back :)

**Comment by fir3pho3nixx** [comment]
> I'd like to contribute in my spare time. Anything I can help with?

@Kralizek - just giving you an idea of what is in play.

**Comment by jonorossi** [comment]
Since we've now merged our netcore branch into master I'm going to close this issue as it'll be in v4.0 which we'll release very soon.

**Comment by fir3pho3nixx** [comment]
Added deprecation notice @ https://github.com/cryosharp/fortress#-deprecation-notice

**Comment by Kralizek** [comment]
Hi everybody! Super happy to see the merge in master branch.

Just a quick question. I already pointed out the limitations for platforms like AWS Lambda to execute .NET Core 1.1 code and marking the package as .NET Standard 1.6 is a step in that direction.

On the other hand, I noticed that the new project file ( https://github.com/castleproject/Windsor/blob/master/src/Castle.Windsor/Castle.Windsor.csproj ) references the 4.3.0 version of many packages. This could lead to indirectly referencing .NET Core 1.1 again.

Would it be possible to downgrade those packages to 4.0 or 4.1?

**Comment by jonorossi** [comment]
@Kralizek could you please create a new issue with the details, we definitely have time to make the change if possible, a pull request would be great if you can. We had to go with .NET Standard 1.6 to use `Microsoft.Extensions.DependencyModel` until we move to .NET Standard 2.0 in the future.

**Comment by alinapopa** [comment]
@Kralizek Castle.Windsor tests are run on netcoreapp1.0
https://github.com/castleproject/Windsor/blob/master/src/Castle.Windsor.Tests/Castle.Windsor.Tests.csproj
Which package is incompatible to .NET Core 1.0 ?

**Comment by fir3pho3nixx** [comment]
@Kralizek - Raised an issue [here](https://github.com/castleproject/Windsor/issues/299)

**Comment by Kralizek** [comment]
Sorry guys for not having replied earlier. I should be able to post a sample sometimes this week.
